ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/Fairness.hs

```diff
@@ -0,0 +1,207 @@
+           <> "exeeds the threshold (" <> show toleranceThreshold <> ")\n"
+           <> "Total number of small transactions that were added: " <> show nrSmall <> "\n"
+           <> "Total number of large transactions that were added: " <> show nrLarge <> "\n"
+           <> "Difference / Total: " <> show diff_ratio
+         )
+
+-- | The fairness test depends on the:
+--
+-- - Total mempool capacity.
+-- - Size of what we consider to be a small transaction.
+-- - Size of what we consider to be a large transaction.
+-- - How many added transactions we count.
+--
+-- The smaller the difference between 'smallTxSize' and 'largeTxSize', the
+-- harder it will be detect potential differences in way the mempool handles
+-- small and large transactions.
+--
+-- The larger the capacity, the higher the chance large transactions can be
+-- added before the mempool is saturated.
+--
+-- We tolerate a certain ratio between the difference of small and large
+-- transactions added, and the total transactions that were added. For instance,
+-- given a threshold of 0.2, if we measure the sizes of 100 added transactions,
+-- the difference between the number small and large transactions we counted
+-- should not be larger than 20.
+--
+data TestParams = TestParams {
+    mempoolMaxCapacity :: Word32
+  , smallTxSize        :: Word32
+  , largeTxSize        :: Word32
+  , nrOftxsToCollect   :: Int
+  , toleranceThreshold :: Double
+  }
+
+-- | Add transactions with with the specified size to the mempool.
+--
+-- We launch a fixed number of adder threads, set to 2 at the moment.
+--
+-- This process does not finish. Hence the 'a' type parameter.
+adders ::
+     TestMempool
+     -- ^ Mempool to which transactions will be added
+  -> Word32
+     -- ^ Transaction size
+  -> IO a
+adders mempool fixedTxSize = Async.mapConcurrently_ adder [0..2] >> error "This process should not finish!"
+  where
+    adder :: Int -> IO ()
+    adder _i = forever $ do
+        -- We pick a random number for the transaction id.
+        thisTxId <- randomIO
+        void $ Mempool.addTxs mempool [mkGenTx thisTxId fixedTxSize]
+
+-- | Remove the given number of transactions and return them.
+--
+-- POSTCONDITION: pure (Just n) = fmap (fmap lenght) $ remover mempool n
+remover ::
+     TestMempool
+     -- ^ Mempool to remove transactions from.
+  -> Int
+  -- ^ Number of transactions to remove.
+  -> IO [Tx]
+remover mempool total = loop [] total
+  where
+    -- Remove transactions one by one till we reach 'n'.
+    loop txs 0 = pure txs
+    loop txs n = do
+        -- We wait 1ms to give the other threads the possibility to add transactions.
+        threadDelay 1000
+        gtx <- atomically $ getOldestTxInMempool
+        -- After we exit the atomic block, tx might not be the oldest
+        -- transaction in the mempool. That is ok, since it is only important
+        -- that we remove __a__ transaction from the mempool.
+        Mempool.removeTxs mempool [Mempool.txId gtx]
```

Comment:
I think it is important to have in mind that this _is not_ how transactions are removed from the mempool. Transactions are removed by becoming invalid either because:
- they are not applicable in the new synced ledger state at all because it is a different fork
- or they are not applicable in the new synced ledger state because they were accepted in a previous ledger state (the new tip usually)

`Mempool.removeTxs` is only called when we forged a block and it could not be adopted by the ChainDB, which indicates that we validated something that the ChainDB rejected, therefore we just clear from the mempool all transactions in that block